#11708 Test cfreactor on macOS on CI

exarkun opened the pull request:
Fixes #11708

exarkun committed:
news fragment

exarkun committed:
tell tox to install macos deps

I guess

exarkun committed:
Merge branch 'trunk' into 11708-macos-cfreactor-ci

glyph committed:
dump flag state if we exist the loop unexpectedly

glyph committed:
better failure

glyph committed:
please, please stop failing codecov

adiroiban commented on codecov.yml:
```diff
@@ -6,6 +6,7 @@
 # the coverage.
 #
 codecov:
+  token: fb52015a-b78e-4d89-a6e0-a4829aea2efa
```
we should not require to add a token for open source project.

I think that the codecov token errors are due to other root causes.

codecov was working without a token for a long time.

glyph replied:
According to the documentation, it's not required.

In practice, there's a race condition on codecov's side, some buggy logic which seems to disproportionately affect projects with larger numbers of github actions statuses that report coverage.

Before we added this coverage was failing something like 40% of builds on this PR; after adding it, I haven't seen one.

adiroiban commented on src/twisted/internet/cfreactor.py:
```diff
@@ -385,6 +385,13 @@ def mainLoop(self):
         finally:
             self._inCFLoop = False
 
+        assert not self.running, (
```
a comment might help us understand why we need this assertion here... or maybe have a different assertion message that can serve to document this code.

glyph replied:
This is part of the WIP and has been removed in the final version.

adiroiban commented:
is this ready for review... it was a draft and then marked as not a draft, so the robot automatically tagged it as ready for review.

i see the macOS tests still failing

glyph committed:
add a test for crash() reliability

exarkun commented:
Sorry, this is still in development.

glyph committed:
report errors eagerly

glyph committed:
remove unused variable, document running check

glyph committed:
use the correct flag

glyph committed:
we need to force the simulation to run at kickoff

glyph committed:
fix non-functional iterate

glyph committed:
debugging

glyph committed:
more logs

glyph committed:
more debugging

glyph committed:
we should be using the current loop and not the main loop

probably doesn't make a difference but since we're dealing with
hard-to-reproduce flakiness let's be as scrupulously correct as possible

glyph committed:
wip

glyph committed:
where do these keep going

glyph committed:
this appears to be the magical combination of fixes

glyph committed:
even more tracing

glyph committed:
clean up logs and remove unnecessary changes

glyph committed:
big, big comments explaining this edge case

glyph committed:
remove now-uninteresting (and syntactically invalid on 3.7) assert

glyph committed:
remove --rterrors addition (let's do this separately)

glyph committed:
remove addSystemEventTrigger change (let's do this separately)

glyph committed:
remove codecov token (let's do this separately)

glyph committed:
love a good sleep()-and-threads test

glyph committed:
this is better failure reporting but it adds coverage problems

tbd: add back the good failure behavior, but put it in a separate
assertRaisesLater utility or something

glyph committed:
replace uncovered lambda: Nones

glyph commented:
please review

adiroiban commented on src/twisted/internet/cfreactor.py:
```diff
@@ -459,7 +517,8 @@ def iterate(self, delay=0):
         """
-        self.callLater(delay, self._stopNow)
+        self._started = True
+        self.callLater(0, self.crash)
```
is the `delay` ignored on purpose? 
... and why do we call crash on `itearate` ?

I think that this need more comments :)

exarkun replied:
2be0fc7aca

adiroiban commented on .github/workflows/test.yaml:
```diff
@@ -129,6 +129,7 @@ jobs:
           - python-version: '3.11'
             runs-on: 'macos-12'
             job-name: 'macos-12-default-tests'
+            tox-env: 'macos-withcov-alldeps'
```
with this change, we make sure that cfreactor deps are installed and we don't skip the cfreactor targeted unit/integration tests, while all the other tests are executed with the select reactor.

If we care about cfreactor, I think is best to run the full test suite with the cfreator...
Most probably it will not detect any other failure, but I think is best to have it and increase the probability of detecting regressions or cfreator-related issues.

and I guess that we still want to run select reactor on macOS

So maybe have these CI macOS environments:

* Python 3.7 with select reactor on macOS 12 - the minimum python that we support
* Python 3.11 with cfreactor reactor on macOS 12 - the maximum/latest python that we support

exarkun replied:
I don't necessarily object but I'm also not going to volunteer for this work.  As-is, this PR adds a lot of CFReactor coverage on CI (it was 0 before, now it's "a lot").  Since you mentioned below that you don't think this is a blocker, I will leave it for another time.

glyph replied:
For reasons related to the way global state works in the corefoundation API, this would be *extremely* difficult to implement without completely rewriting trial from scratch. So I am not sure we should even bother tackling this later.

glyph replied:
(In any case, definitely nothing to do on this PR so resolving)

adiroiban commented:
I left just a few comments as I am not familiar with macOS or cfreactor.
The CI comment is not a blocker.
The  `iterate` part looks strange. I was expecting the tests to fail with such a change.

exarkun committed:
comment about the differently-shaped iterate implementation

exarkun commented on src/twisted/internet/cfreactor.py:
```diff
@@ -132,7 +132,7 @@ def __init__(self, runLoop=None, runner=None):
 
         if runLoop is None:
-            runLoop = CFRunLoopGetMain()
+            runLoop = CFRunLoopGetCurrent()
```
`CFRunLoopGetMain` gets "the" main loop.  I suppose this is "the main thread's" event loop?
`CFRunLoopGetCurrent` gets "the current thread's" main loop.

The latter seems more likely to be correct in general.

* https://developer.apple.com/documentation/corefoundation/1542428-cfrunloopgetcurrent 
* https://developer.apple.com/documentation/corefoundation/1542890-cfrunloopgetmain

Both doc pages say the API follows "The Get Rule".  Do the Python bindings deal with that for us or are we supposed to be making `CFRetain` calls from Python (@glyph )?

glyph replied:
Correct on both counts. It's pyobjc's job, generally, to deal with retention rules.

exarkun commented on src/twisted/internet/cfreactor.py:
```diff
@@ -372,22 +372,85 @@ def _moveCallLaterSooner(self, tple):
+        super().startRunning(installSignalHandlers)
+        # We must force the simulator to run because in addition to scheduling
+        # timed calls, super().startRunning also fires startup events which may
+        # crash the reactor.
```
I don't quite follow this explanation.  I tried to expand some of the implications here but I don't think I got somewhere coherent:

```
        # Prior to this point (maybe because of the super().startRunning)                                                                                   
        # arbitrary application code may have run.  This may include code that                                                                              
        # added event sources (timers or file descriptors) to the reactor.  If                                                                              
        # so, either _scheduleSimulate will already have run and added a Timer                                                                              
        # to the CFRunLoop or _watchFD will have added a Source.  Either way,                                                                               
        # there is work for CFRunLoop to do and things will work well.                                                                                      
        #                                                                                                                                                   
        # If neither of those things happened then there is no work for the                                                                                 
        # CFRunLoop to do.                                                                                                                                  
        #                                                                                                                                                   
        # ... so what?                                                                                                                                      
```

glyph replied:
This is … directionally correct, so I'll try to do an update that addresses some of these problems

glyph replied:
Aaand upon improving the verbiage I realized the previous version was wrong anyway. Oops! Hopefully the new version is correct.

exarkun commented on src/twisted/internet/cfreactor.py:
```diff
@@ -372,22 +372,85 @@ def _moveCallLaterSooner(self, tple):
         the run loop until C{crash()} is called.
         """
-        self._inCFLoop = True
+        # we were crashed during startup.
+        if not self._started:
+
```
I'd put this comment inside the `if` since that's the case it's describing.  Also I tried to expand this one and didn't succeed:

```
            # startRunning set _started to True so between then and now,                                                                                                          
            # something crashed us.  There aren't a lot of candidates for                                                                                                         
            # this.  The most well-known is:                                                                                                                                      
            #                                                                                                                                                                     
            #   reactor.callWhenRunning(reactor.crash)                                                                                                                            
            #                                                                                                                                                                     
            # We *don't* eagerly return here because ...                                                                                                                          
            #                                                                                                                                                                     
            #    why?                                                                                                                                                             
            #                                                                                                                                                                     
            # To avoid that, arrange to let the CF main loop run very briefly.                                                                                                    
```

glyph replied:
OK I think that should do it

exarkun commented on src/twisted/internet/cfreactor.py:
```diff
@@ -372,22 +372,85 @@ def _moveCallLaterSooner(self, tple):
+                self.crash()
+
+            # we're running again, just for a moment, while we wait for some
+            # I/O to take place.  We need this to be True for callLater to not
```
The "I/O" that we wait for is just the CF Timer to fire and call our crash function right?

glyph replied:
What I was trying to get across here is that we need to call into the actual loop, not just run the logic of the timer immediately. To allow some chance for already-ready file descriptors to be invoked. So the "I/O" is explicitly in contrast to this logic.

I'll try to make this clearer

exarkun commented on src/twisted/internet/cfreactor.py:
```diff
@@ -372,22 +372,85 @@ def _moveCallLaterSooner(self, tple):
+                    # that in cases where the reactor is being rapidly started
+                    # and crashed repeatedly, and sources are being rapidly
+                    # invalidated both inside and outside the loop (e.g. in the
+                    # Twisted test suite), CoreFoundation (possibly
+                    # incorrectly?) exits the runloop early.  To accommodate
+                    # this stress case, we make extra sure that our scheduled
+                    # timer is re-created on the loop as a CFRunLoopTimer.
```
```suggestion
                    # Twisted test suite), CoreFoundation exits the runloop even
                    # when there *is* some work to do. To accommodate this
                    # stress case, we make extra sure that our scheduled
                    # timer is re-created on the loop as a CFRunLoopTimer.
```

glyph replied:
I cannot prove that there is in fact "work to do", at least according to the semantics established by CFRunLoop: there are various points where sources are automatically invalidated (when timers expire, I think), and some of our uses of the API (for example, `CFSocketCreateWithNative` with a pipe file descriptor rather than a socket) are taking some liberties. So it's possible that we just haven't correctly appreciated some nuance of automatic invalidation

glyph replied:
Much lengthier comment hopefully explains the situation in detail.

exarkun commented on src/twisted/internet/test/test_cfreactor.py:
```diff
@@ -0,0 +1,99 @@
+            stopped.append("yes")
+
+        def sleepThenStop() -> None:
+            sleep(0.5)
```
What's the point of the sleep?  What order of events is this test trying to exercise?

glyph replied:
The goal here is to start the loop with no timers, so that scheduleSimulate exercises the "timeout is None" case. If we don't sleep, then the shutdown runs scheduleSimulate with force=True, forcing an immediate shutdown and no further calls to scheduleSimulate. We have to run through startup while fully running the loop, then have some I/O (waker.doRead) wake us up with no timers in sight.

glyph replied:
Maybe I'm wrong about this. I tried to explain it to myself, and I could not construct the sequence of events that I thought it was there to illustrate.  If codecov agrees then I think we can drop it.

glyph replied:
OK. I was wrong. The coverage gap I saw on that line must have been for unrelated reasons. The sleep is gone!

glyph committed:
address feedback on monster comments by making them more monstrous

glyph committed:
try to be clearer about our level of certainty about this behavior

glyph committed:
remove potentially unnecessary sleep

after trying for like an hour to explain the need for this I convinced myself
it's not necessary. so let's see if codecov agrees with me

glyph commented:
OK, I think that this is pretty good to land now. Does anyone feel like an approval?

adiroiban commented on src/twisted/internet/test/test_cfreactor.py:
```diff
@@ -0,0 +1,97 @@
+    """
+
+
+noop()
```
If this is needed, I think that it needs a comment.
It is not obvious why this is needed.

adiroiban commented on src/twisted/internet/test/test_cfreactor.py:
```diff
@@ -0,0 +1,97 @@
+
+    def test_whiteboxIterate(self) -> None:
+        """
+        C{.iterate()} starts the main loop, then crashes the reactor.
```
I don't understand this test.
My understanding is that the `crash` part of iterate is a private implementation detail.

For the tests, I was expecting not to have it mentioned in relation to iterate.

----------

So maybe the description could be

`iterate() executes any callLater callbacks and leaves the reactor stopped`

I guess that the purpose of this test is to explain that iterate on cfreactor has a different behaviour than other reactors.

So maybe, in this test we should have `r.iterate(4)` and it should explain that even if requested to run for 4 seconds it will also execute the delayed called scheduled for 100 seconds.

glyph replied:
The exiting, black-box tests already test this, but I'll update the docstring to explain the internal state being tested.

adiroiban commented on src/twisted/internet/cfreactor.py:
```diff
@@ -401,24 +500,28 @@ def _scheduleSimulate(self, force=False):
+            return
+
+        timeout = 0.0 if force else self.timeout()
+        if timeout is None:
```
I know that this was already in trunk.
I is not clear what is the semantic of `timeout = None`

glyph replied:
This is a common feature of pretty much every reactor; it just means "there are no scheduled calls". 

This ought to be documented in `twisted.internet.base.ReactorBase.timeout` but that's a separate change :)

adiroiban reviewed (approved):
Thanks for your work !

I left a few minor comments.
I see that the changes from this PR are done by both Jean Paul and Glyph.

If you are happy, I think that this should be merged.

I have never used the cfreactor... so I am not the best person to approve, but I am not sure if anyone else will have  time to review it soon.

And if  anyone cares about cfreactor, I think we can always follow up with a new PR.

My understanding is that the code from trunk is broken anyway.

Regards

glyph committed:
address feedback: explain why noop is called

glyph committed:
address feedback: explain what test_whiteboxIterate is verifying

glyph commented:
> My understanding is that the code from trunk is broken anyway.

Technically this is accurate, but "broken" as a blanket term is misleading. The reactor before this change works fine for applications, and even *mostly* passes the Twisted suite itself.  There are just some edge case race conditions that specifically get triggered in the very unusual context of Twisted's tests, where thousands of instances of the CF reactor are being instantiated and started and stopped.

While the implementation is now more correct, I do not expect that its correctness will be visible to applications, which is why I decided to go with a `.misc` change and not include the robustness improvements in the changelog; the robustness is mostly just for our own usage in our own testing.

The problem here is that CoreFoundation provides tons of global, un-inspectable state, which we cannot test against and we cannot isolate to individual instances, only to threads.
